sharkfuser/include/fusili/asm_emitter.h

```diff
@@ -0,0 +1,420 @@
+// to disambiguate from the `{}` that `std::format` uses for replacements.
+// If not you'll hit a compilation error like so:
+//    "error: call to consteval function 'std::basic_format_string<char, ...'"
+//    "is not a constant expression"
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef FUSILI_ASM_EMITTER_H
+#define FUSILI_ASM_EMITTER_H
+
+#include "fusili/attributes/tensor_attributes.h"
+#include "fusili/graph.h"
+#include "fusili/node/conv_node.h"
+#include "fusili/types.h"
+
+#include <cassert>
+#include <format>
+#include <sstream>
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace fusili {
+
+// Map from Fusili types to MLIR types.
+static const std::unordered_map<DataType, std::string> DataTypeToMlirTypeAsm = {
+    {DataType::Half, "f16"},       {DataType::BFloat16, "bf16"},
+    {DataType::Float, "f32"},      {DataType::Double, "f64"},
+    {DataType::Uint8, "ui8"},      {DataType::Int8, "si8"},
+    {DataType::Int16, "si16"},     {DataType::Int32, "si32"},
+    {DataType::Int64, "si64"},     {DataType::Boolean, "i1"},
+    {DataType::FP8E5M2, "f8E5M2"},
+};
+
+// Given a vector of ints, returns the MLIR assembly for the
+// `torch.constant.int` ops for each int value and the
+// `torch.prim.ListConstruct` op wrapping these into a single
+// value.
+//
+// For example if `getListOfIntOpsAsm` is called on these inputs:
+//    listOfInts: {1, 2}
+//    prefix: "stride"
+//    suffix: "conv"
+//
+// It generates the following MLIR assembly:
+//
+//   %stride_val_0_conv = torch.constant.int 1
+//   %stride_val_1_conv = torch.constant.int 2
+//   %stride_conv = torch.prim.ListConstruct
+//          %stride_val_0_conv, %stride_val_1_conv :
+//              (!torch.int, !torch.int) -> !torch.list<int>
+//
+// The prefix is generally what attribute this refers to (e.g.
+// padding, stride, dilation etc.) and the suffix is the node's
+// unique name (for SSA disambiguation).
+inline std::string getListOfIntOpsAsm(const std::vector<int64_t> &listOfInts,
+                                      const std::string &prefix,
+                                      const std::string &suffix) {
+  std::ostringstream oss;
+  std::vector<std::string> ssaValueNames;
+
+  // Emit `torch.constant.int` ops for each int value
+  for (size_t i = 0; i < listOfInts.size(); ++i) {
+    std::string ssaValueName =
+        "%" + prefix + "_val_" + std::to_string(i) + "_" + suffix;
+    oss << ssaValueName << " = torch.constant.int " << listOfInts[i]
+        << "\n    ";
+    ssaValueNames.push_back(ssaValueName);
+  }
+
+  // Emit the ListConstruct op
+  oss << "%" + prefix + "_" + suffix << " = torch.prim.ListConstruct ";
+  for (size_t i = 0; i < ssaValueNames.size(); ++i) {
+    if (i > 0)
+      oss << ", ";
+    oss << ssaValueNames[i];
+  }
+  oss << " : (";
+  for (size_t i = 0; i < ssaValueNames.size(); ++i) {
+    if (i > 0)
+      oss << ", ";
+    oss << "!torch.int";
+  }
```

Comment:
You might want a utility function similar to LLVM's [interleave](https://github.com/llvm/llvm-project/blob/028f70d2524be56bb5d191ec3d7022b40bc4859b/llvm/include/llvm/ADT/STLExtras.h#L2167-L2192).

Reply:
Nice! Done.

Reply:
I went ahead and added an overload of `interleave` as well, which accepts a `SkipFunctor` to conditionally skip elements based on a predicate. This is useful for cases where I needed to check `input->isScalar()` or `output->isVirtual()` and skip those from the result.